gdx/src/com/badlogic/gdx/graphics/g2d/TextureArraySpriteBatch.java

```diff
@@ -0,0 +1,1435 @@
+			&& blendDstFuncAlpha == dstFuncAlpha) {
+			return;
+		}
+
+		flush();
+
+		blendSrcFunc = srcFuncColor;
+		blendDstFunc = dstFuncColor;
+		blendSrcFuncAlpha = srcFuncAlpha;
+		blendDstFuncAlpha = dstFuncAlpha;
+	}
+
+	@Override
+	public int getBlendSrcFunc () {
+		return blendSrcFunc;
+	}
+
+	@Override
+	public int getBlendDstFunc () {
+		return blendDstFunc;
+	}
+
+	@Override
+	public int getBlendSrcFuncAlpha () {
+		return blendSrcFuncAlpha;
+	}
+
+	@Override
+	public int getBlendDstFuncAlpha () {
+		return blendDstFuncAlpha;
+	}
+
+	@Override
+	public boolean isBlendingEnabled () {
+		return !blendingDisabled;
+	}
+
+	@Override
+	public boolean isDrawing () {
+		return drawing;
+	}
+
+	@Override
+	public Matrix4 getProjectionMatrix () {
+		return projectionMatrix;
+	}
+
+	@Override
+	public Matrix4 getTransformMatrix () {
+		return transformMatrix;
+	}
+
+	@Override
+	public void setProjectionMatrix (Matrix4 projection) {
+
+		if (drawing) {
+			flush();
+		}
+
+		projectionMatrix.set(projection);
+
+		if (drawing) {
+			setupMatrices();
+		}
+	}
+
+	@Override
+	public void setTransformMatrix (Matrix4 transform) {
+
+		if (drawing) {
+			flush();
+		}
+
+		transformMatrix.set(transform);
+
+		if (drawing) {
+			setupMatrices();
+		}
+	}
+
+	private void setupMatrices () {
+
+		combinedMatrix.set(projectionMatrix).mul(transformMatrix);
+
+		if (customShader != null) {
+			customShader.setUniformMatrix("u_projTrans", combinedMatrix);
+			Gdx.gl20.glUniform1iv(customShader.fetchUniformLocation("u_textures", true), maxTextureUnits, textureUnitIndicesBuffer);
+
+		} else {
+			shader.setUniformMatrix("u_projTrans", combinedMatrix);
+			Gdx.gl20.glUniform1iv(shader.fetchUniformLocation("u_textures", true), maxTextureUnits, textureUnitIndicesBuffer);
+		}
+	}
+
+	/** Queries the number of supported textures in a texture array by trying the create the default shader.<br>
+	 * The first call of this method is very expensive, after that it simply returns a cached value.
+	 * @return the number of supported textures in a texture array or zero if this feature is unsupported on this device.
+	 * @see {@link #setShader(ShaderProgram shader)} */
+	public static int getMaxTextureUnits () {
+
+		if (maxTextureUnits == -1) {
+
+			// Query the number of available texture units and decide on a safe number of texture units to use
+			IntBuffer texUnitsQueryBuffer = BufferUtils.newIntBuffer(32);
+
+			Gdx.gl.glGetIntegerv(GL20.GL_MAX_TEXTURE_IMAGE_UNITS, texUnitsQueryBuffer);
+
+			int maxTextureUnitsLocal = texUnitsQueryBuffer.get();
+
+			// Some OpenGL drivers (I'm looking at you, Intel!) do not report the right values,
+			// so we take caution and test it first, reducing the number of slots if needed.
+			// Will try to find the maximum amount of texture units supported.
+			while (maxTextureUnitsLocal > 0) {
+
+				try {
+
+					createDefaultShader(maxTextureUnitsLocal);

```

Comment:
I think we should dispose the shader if it compiled and you are not using this one